src/librustc/hir/lowering.rs

```diff
@@ -1149,13 +1149,69 @@ impl<'a> LoweringContext<'a> {
         }
     }
 
-    fn lower_generic_arg(&mut self,
-                         arg: &ast::GenericArg,
-                         itctx: ImplTraitContext<'_>)
-                         -> hir::GenericArg {
+    fn lower_generic_arg(
+        &mut self,
+        arg: &ast::GenericArg,
+        itctx: ImplTraitContext<'_>
+    ) -> hir::GenericArg {
         match arg {
             ast::GenericArg::Lifetime(lt) => GenericArg::Lifetime(self.lower_lifetime(&lt)),
-            ast::GenericArg::Type(ty) => GenericArg::Type(self.lower_ty_direct(&ty, itctx)),
+            ast::GenericArg::Type(ty) => {
+                // A path type argument could be an incorrectly parsed generic const argument. If
+                // we resolved the path in the value namespace, we transform it into a generic
+                // const argument.
+                if let TyKind::Path(ref qself, ref path) = ty.kind {
+                    if path.segments.len() == 1 {
+                        if let Some(partial_res) = self.resolver.get_partial_res(ty.id) {
+                            match partial_res.base_res() {
+                                Res::Err | Res::Def(DefKind::Ctor(_, CtorKind::Fictive), ..) => (),
+                                res if res.is_valid_in_ns(Namespace::ValueNS) => {
+                                    debug!(
+                                        "lower_generic_arg: Lowering incorrectly parsed type \
+                                        argument as const argument: {:?}",
+                                        ty,
+                                    );
+
+                                    // Construct a AnonConst where the expr is the "ty"'s path.
+
+                                    let parent_def_index =
+                                        self.current_hir_id_owner.last().unwrap().0;
+                                    let node_id = self.sess.next_node_id();
+
+                                    // Add a definition for the in-band const def.
+                                    self.resolver.definitions().create_def_with_parent(
+                                        parent_def_index,
+                                        node_id,
+                                        DefPathData::AnonConst,
+                                        ExpnId::root(),
+                                        ty.span,
+                                    );
```

Comment:
Hmm, this distinction between types and consts (consts requiring an extra def path node) is something I didn't expect.
@eddyb, is it possible that creating this node during lowering is too late?

Reply:
Maybe? But if you limit this to a really simple path with no generic args, there shouldn't be any children defs nested inside so making a leaf `AnonConst` should be fine.

Reply:
@yodaldevoid 
Could you bring back the "no generic args" restriction then?
(It should rarely be relevant in practice.)

Reply:
It is done.

It should be noted that this does mean that we cannot remove all of the braces from the const-generics tests, specifically in "fn-const-param-infer.rs" which uses type params all over the place. How hard/wise would it be to iterate through the children of the new def to create defs for them if needed?

Reply:
>How hard/wise would it be to iterate through the children of the new def to create defs for them if needed?

I'm not sure.
First, we need to at least construct a test where things would break due to def paths not being constructed properly.
This is a material for another PR, IMO.
We get, like, 90% of benefits from disambiguating simple identifiers, so it would be nice to land this PR sooner.

Reply:
I think we already have such a test in "fn-const-param-infer.rs", but I agree that is something for a future PR.